## Problem Description

The goal of this linear programming problem is to determine the number of servings of each food item to consume per week in order to meet the minimum weekly nutritional requirements while minimizing the total cost of these food items. The five food items under consideration are Eggs, Hash Browns, Chobani Yogurt, Safeway Cobb Salad, and Canned Tuna. Each food item has a specific cost and provides varying amounts of nutrients such as calories, protein, vitamin D, calcium, iron, potassium, and sodium.

The decision variables in this problem are the number of servings per week for each of the food items, which we'll represent as:

- **x1** = number of servings of Eggs per week
- **x2** = number of servings of Hash Browns per week
- **x3** = number of servings of Yogurt per week
- **x4** = number of servings of Cobb Salad per week
- **x5** = number of servings of Canned Tuna per week

The **objective** of this problem is to minimize the total cost of the food consumed over the week, represented by the sum of the cost per serving of each food item multiplied by the number of servings consumed.

The **nutritional constraints** ensure that the minimum weekly intake for calories, protein, vitamin D, calcium, iron, and potassium is met, and that the sodium intake does not exceed the maximum allowed level. These constraints are as follows:

- Total calories consumed must be at least 14,000 kcal (equivalent to 2,000 kcal per day for 7 days).
- Total protein consumed must be at least 350 grams (equivalent to 50 grams per day for 7 days).
- Total vitamin D consumed must be at least 140 micrograms.
- Total calcium consumed must be at least 9,100 milligrams.
- Total iron consumed must be at least 126 milligrams.
- Total potassium consumed must be at least 32,900 milligrams.
- Total sodium consumed should not exceed 35,000 milligrams (5,000 mg per day).

These constraints ensure that a balanced diet is maintained over the week while minimizing costs.

---

# Part 2: Problem in Standard Form

### Decision Variables

- **x1** (Eggs), **x2** (Hash Browns), **x3** (Yogurt), **x4** (Cobb Salad), **x5** (Canned Tuna) are the servings per week of each food item, with the constraint that \( x_i \geq 0 \) for all \( i \).

### Objective Function (Minimize total cost)

\[
\text{Minimize } 0.50x_1 + 0.28x_2 + 1.50x_3 + 4.49x_4 + 1.39x_5
\]

### Nutritional Constraints

1. **Calories**: 
\[
70x_1 + 120x_2 + 140x_3 + 250x_4 + 100x_5 \geq 14000
\]
2. **Protein**: 
\[
6x_1 + 1x_2 + 5x_3 + 12x_4 + 20x_5 \geq 350
\]
3. **Vitamin D**: 
\[
1x_1 + 0x_2 + 3x_3 + 0x_4 + 1x_5 \geq 140
\]
4. **Calcium**: 
\[
28x_1 + 10x_2 + 190x_3 + 150x_4 + 10x_5 \geq 9100
\]
5. **Iron**: 
\[
1x_1 + 0.4x_2 + 0x_3 + 0.48x_4 + 1x_5 \geq 126
\]
6. **Potassium**: 
\[
69x_1 + 210x_2 + 260x_3 + 200x_4 + 200x_5 \geq 32900
\]
7. **Sodium**: 
\[
65x_1 + 250x_2 + 80x_3 + 860x_4 + 250x_5 \leq 35000
\]

# Part 3: Solving the linear programming problem

In [1]:
import pulp

In [12]:
# Note: I chose to create a function to solve the diet problem with the option to set a minimum number of servings for each food item. This will make it re-usable for part 4 without having to rewrite the function.

def solve_diet_problem(min_servings=0, filename=None):
    # Create the LP problem instance
    prob = pulp.LpProblem("The_Diet_Problem", pulp.LpMinimize)

    # Decision variables: Servings per week of each food item
    # Set the lower bound to min_servings (default is 0)
    x1 = pulp.LpVariable('Eggs', lowBound=min_servings)
    x2 = pulp.LpVariable('HashBrowns', lowBound=min_servings)
    x3 = pulp.LpVariable('Yogurt', lowBound=min_servings)
    x4 = pulp.LpVariable('CobbSalad', lowBound=min_servings)
    x5 = pulp.LpVariable('CannedTuna', lowBound=min_servings)

    # Cost per serving of each food item
    cost = {
        x1: 0.50,   # Eggs
        x2: 0.28,   # Hash Browns
        x3: 1.50,   # Chobani Yogurt Strawberry Banana Low Fat Greek Yogurt Drink (7 fl oz)
        x4: 4.49,   # Safeway Cobb Salad
        x5: 1.50    # StarKist Chunk Light Tuna in Oil
    }

    # Objective function: Minimize total cost over the week
    prob += pulp.lpSum([cost[var] * var for var in [x1, x2, x3, x4, x5]]), "Total Cost"

    # Nutrient content per serving
    calories = {
        x1: 70,
        x2: 120,
        x3: 140,
        x4: 210,
        x5: 100
    }

    protein = {
        x1: 6,
        x2: 1,
        x3: 10,
        x4: 11,
        x5: 16
    }

    vitamin_d = {
        x1: 1,
        x2: 0,
        x3: 0,
        x4: 0.8,   # Converted 4% of 20 mcg to mcg
        x5: 2.3
    }

    calcium = {
        x1: 28,
        x2: 10,
        x3: 220,
        x4: 52,    # Converted 4% of 1300 mg to mg
        x5: 0
    }

    iron = {
        x1: 1,
        x2: 0.4,
        x3: 0,
        x4: 1.08,  # Converted 6% of 18 mg to mg
        x5: 0.9
    }

    potassium = {
        x1: 69,
        x2: 210,
        x3: 340,
        x4: 282,   # Converted 6% of 4700 mg to mg
        x5: 170
    }

    sodium = {
        x1: 65,
        x2: 250,
        x3: 90,
        x4: 760,
        x5: 320
    }

    # Weekly nutritional requirements
    calories_min = 14000  # 2000 kcal/day * 7 days
    protein_min = 350     # 50 g/day * 7 days
    vitamin_d_min = 140   # 20 mcg/day * 7 days
    calcium_min = 9100    # 1300 mg/day * 7 days
    iron_min = 126        # 18 mg/day * 7 days
    potassium_min = 32900 # 4700 mg/day * 7 days
    sodium_max = 35000    # 5000 mg/day * 7 days

    # Nutritional constraints
    # Calories constraint
    prob += pulp.lpSum([calories[var] * var for var in [x1, x2, x3, x4, x5]]) >= calories_min, "CaloriesRequirement"

    # Protein constraint
    prob += pulp.lpSum([protein[var] * var for var in [x1, x2, x3, x4, x5]]) >= protein_min, "ProteinRequirement"

    # Vitamin D constraint
    prob += pulp.lpSum([vitamin_d[var] * var for var in [x1, x2, x3, x4, x5]]) >= vitamin_d_min, "VitaminDRequirement"

    # Calcium constraint
    prob += pulp.lpSum([calcium[var] * var for var in [x1, x2, x3, x4, x5]]) >= calcium_min, "CalciumRequirement"

    # Iron constraint
    prob += pulp.lpSum([iron[var] * var for var in [x1, x2, x3, x4, x5]]) >= iron_min, "IronRequirement"

    # Potassium constraint
    prob += pulp.lpSum([potassium[var] * var for var in [x1, x2, x3, x4, x5]]) >= potassium_min, "PotassiumRequirement"

    # Sodium constraint (maximum)
    prob += pulp.lpSum([sodium[var] * var for var in [x1, x2, x3, x4, x5]]) <= sodium_max, "SodiumLimit"

    # Solve the problem (suppressing solver output)
    prob.solve(pulp.PULP_CBC_CMD(msg=False))

    # Check the status of the solution
    status = pulp.LpStatus[prob.status]
    print("Status:", status)

    if status == 'Optimal':
        # Print the optimal servings of each food item
        print("Optimal servings per week:")
        for var in [x1, x2, x3, x4, x5]:
            print(f"{var.name} = {var.varValue:.2f} servings")

        # Print the total minimum cost
        total_cost = pulp.value(prob.objective)
        print(f"\nTotal Cost = ${total_cost:.2f} per week")

        # Calculate total nutrient intake
        total_calories = sum([calories[var] * var.varValue for var in [x1, x2, x3, x4, x5]])
        total_protein = sum([protein[var] * var.varValue for var in [x1, x2, x3, x4, x5]])
        total_vitamin_d = sum([vitamin_d[var] * var.varValue for var in [x1, x2, x3, x4, x5]])
        total_calcium = sum([calcium[var] * var.varValue for var in [x1, x2, x3, x4, x5]])
        total_iron = sum([iron[var] * var.varValue for var in [x1, x2, x3, x4, x5]])
        total_potassium = sum([potassium[var] * var.varValue for var in [x1, x2, x3, x4, x5]])
        total_sodium = sum([sodium[var] * var.varValue for var in [x1, x2, x3, x4, x5]])

        print("\nTotal Nutrient Intake Over the Week:")
        print(f"Calories: {total_calories:.2f} kcal (Weekly Requirement: {calories_min} kcal)")
        print(f"Protein: {total_protein:.2f} g (Weekly Requirement: {protein_min} g)")
        print(f"Vitamin D: {total_vitamin_d:.2f} mcg (Weekly Requirement: {vitamin_d_min} mcg)")
        print(f"Calcium: {total_calcium:.2f} mg (Weekly Requirement: {calcium_min} mg)")
        print(f"Iron: {total_iron:.2f} mg (Weekly Requirement: {iron_min} mg)")
        print(f"Potassium: {total_potassium:.2f} mg (Weekly Requirement: {potassium_min} mg)")
        print(f"Sodium: {total_sodium:.2f} mg (Weekly Limit: {sodium_max} mg)")

        if filename:
            # Save the results to a text file
            with open(filename, 'w') as file:
                file.write("Optimal Servings Per Week:\n")
                for var in [x1, x2, x3, x4, x5]:
                    file.write(f"{var.name} = {var.varValue:.2f} servings\n")

                file.write(f"\nTotal Cost = ${total_cost:.2f} per week\n\n")
                file.write("Total Nutrient Intake Over the Week:\n")
                file.write(f"Calories: {total_calories:.2f} kcal (Weekly Requirement: {calories_min} kcal)\n")
                file.write(f"Protein: {total_protein:.2f} g (Weekly Requirement: {protein_min} g)\n")
                file.write(f"Vitamin D: {total_vitamin_d:.2f} mcg (Weekly Requirement: {vitamin_d_min} mcg)\n")
                file.write(f"Calcium: {total_calcium:.2f} mg (Weekly Requirement: {calcium_min} mg)\n")
                file.write(f"Iron: {total_iron:.2f} mg (Weekly Requirement: {iron_min} mg)\n")
                file.write(f"Potassium: {total_potassium:.2f} mg (Weekly Requirement: {potassium_min} mg)\n")
                file.write(f"Sodium: {total_sodium:.2f} mg (Weekly Limit: {sodium_max} mg)\n")
            
            print(f"\nResults saved to '{filename}'")

        # Return the results as a dictionary
        results = {
            'Status': status,
            'Total Cost': total_cost,
            'Servings': {var.name: var.varValue for var in [x1, x2, x3, x4, x5]},
            'Total Nutrients': {
                'Calories': total_calories,
                'Protein': total_protein,
                'Vitamin D': total_vitamin_d,
                'Calcium': total_calcium,
                'Iron': total_iron,
                'Potassium': total_potassium,
                'Sodium': total_sodium
            }
        }
        return results
    else:
        print("No optimal solution found.")
        return None

In [11]:
# Run the function to solve the diet problem
results = solve_diet_problem(filename="part_3_results.txt")

Status: Optimal
Optimal servings per week:
Eggs = 140.00 servings
HashBrowns = 78.31 servings
Yogurt = 19.99 servings
CobbSalad = 0.00 servings
CannedTuna = 0.00 servings

Total Cost = $121.91 per week

Total Nutrient Intake Over the Week:
Calories: 21995.05 kcal (Weekly Requirement: 14000 kcal)
Protein: 1118.17 g (Weekly Requirement: 350 g)
Vitamin D: 140.00 mcg (Weekly Requirement: 140 mcg)
Calcium: 9100.00 mg (Weekly Requirement: 9100 mg)
Iron: 171.32 mg (Weekly Requirement: 126 mg)
Potassium: 32900.00 mg (Weekly Requirement: 32900 mg)
Sodium: 30475.84 mg (Weekly Limit: 35000 mg)


## Results Overview

Based on the results, the minmimum cost solution is to consume the following servings per week:
- **Eggs**: 140.00 servings per week
- **Hash Browns**: 78.31 servings per week
- **Yogurt**: 19.99 servings per week
- **Cobb Salad**: 18.00 servings per week
- **Canned Tuna**: 0.00 servings per week

Which results in a cost of $121.91 per week.

# Part 4: Adding variety to linear programming problem

For the modified problem, we will add the constraint that each food item should have at least 1 serving per week. This constraint ensures that each food item is included in the diet to provide variety and a balanced nutritional intake. 

In [14]:
results = solve_diet_problem(min_servings=1, filename="part_4_results.txt")

Status: Optimal
Optimal servings per week:
Eggs = 136.90 servings
HashBrowns = 76.81 servings
Yogurt = 20.21 servings
CobbSalad = 1.00 servings
CannedTuna = 1.00 servings

Total Cost = $126.26 per week

Total Nutrient Intake Over the Week:
Calories: 21939.70 kcal (Weekly Requirement: 14000 kcal)
Protein: 1127.33 g (Weekly Requirement: 350 g)
Vitamin D: 140.00 mcg (Weekly Requirement: 140 mcg)
Calcium: 9100.00 mg (Weekly Requirement: 9100 mg)
Iron: 169.60 mg (Weekly Requirement: 126 mg)
Potassium: 32900.00 mg (Weekly Requirement: 32900 mg)
Sodium: 30999.63 mg (Weekly Limit: 35000 mg)

Results saved to 'part_4_results.txt'


The results of the modified problem show that each food item has at least 1 serving per week. It does not change the results significantly, as it simply enforces a minimum variety in the diet but adding one serving for each that were not previously used in the optimal solution. The total cost increased slightly from 121.91 to 126.26.

The result still lacks in variety. A few suggestions on how I could update this to result in greater diet variety:

- Introduce additional nutritional constraints. For example, a maximum cholesterol intake, maximum sugar intake, minimum fiber intake, etc.

- Add an upper bound for number of servings for an item. This would allow for more variety by limiting the amount of a single food item consumed.

- Introduce a penalty for consuming the same food item multiple times. For example, an additional 'hidden' variable that increases the cost if the same food item is consumed consecutively that could be used in the calculation. This would not reflect in the actual cost but would influence the optimization to prefer variety. 

# Part 5: LLM Usage to solve the Diet Problem

For full transparency, I use ChatGPT for almost all tasks I perform in work and school. My parts 2, 3 and 4 were done in conjunction with ChatGPT models 4o and o1-mini. I used it as a starting point and built around it. Additionally, I use the JetBrains IDE PyCharm with GitHub’s Copilot integrated which helps with my code writing as well.

For this part of the assignment, I chose to use Anthropic’s Claude 3.5 Sonnet (since I pay for the premium models of ChatGPT already). This can be found here: https://claude.ai/

Here were my prompts and output:

___

Prompt:  
Hi, I am currently pursuing my masters in Data Science at Northwestern University and am working on an assignment for my Decision Analytics class. I need a model built to solve the Diet Problem. Please provide a plain text description and parameterized Python code to assist with solving the problem.

Response:  

I'll help you create a model for the Diet Problem, which is a classic optimization problem in operations research. Let me provide you with both a description and a Python implementation using linear programming.

```python
import pulp

def solve_diet_problem(
    foods,           # Dictionary of foods with their costs per unit
    nutrients,       # List of nutrients to consider
    food_nutrients,  # Dictionary of nutrient content per unit of food
    min_nutrients,   # Dictionary of minimum daily requirements for each nutrient
    max_nutrients    # Dictionary of maximum daily allowances for each nutrient
):
    # Create the optimization problem
    prob = pulp.LpProblem("Diet_Optimization", pulp.LpMinimize)
    
    # Decision variables: how much of each food to buy
    food_vars = pulp.LpVariable.dicts("Food", foods.keys(), lowBound=0)
    
    # Objective function: minimize cost
    prob += pulp.lpSum([foods[f] * food_vars[f] for f in foods])
    
    # Constraints for minimum nutrient requirements
    for nutrient in nutrients:
        prob += pulp.lpSum([food_nutrients[f][nutrient] * food_vars[f] for f in foods]) >= min_nutrients[nutrient], f"Min_{nutrient}"
    
    # Constraints for maximum nutrient allowances
    for nutrient in nutrients:
        if nutrient in max_nutrients:
            prob += pulp.lpSum([food_nutrients[f][nutrient] * food_vars[f] for f in foods]) <= max_nutrients[nutrient], f"Max_{nutrient}"
    
    # Solve the problem
    prob.solve()
    
    # Prepare the results
    results = {
        'status': pulp.LpStatus[prob.status],
        'optimal_cost': pulp.value(prob.objective),
        'food_amounts': {f: pulp.value(food_vars[f]) for f in foods},
        'nutrients_obtained': {
            n: sum(food_nutrients[f][n] * pulp.value(food_vars[f]) for f in foods)
            for n in nutrients
        }
    }
    
    return results

# Example usage
if __name__ == "__main__":
    # Sample data
    foods = {
        'chicken': 2.89,    # price per unit
        'beef': 3.84,
        'fish': 2.63,
        'eggs': 1.45,
        'beans': 0.95
    }
    
    nutrients = ['protein', 'fat', 'carbs']
    
    food_nutrients = {
        'chicken': {'protein': 27, 'fat': 14, 'carbs': 0},
        'beef': {'protein': 26, 'fat': 19, 'carbs': 0},
        'fish': {'protein': 22, 'fat': 12, 'carbs': 0},
        'eggs': {'protein': 13, 'fat': 10, 'carbs': 1},
        'beans': {'protein': 15, 'fat': 1, 'carbs': 40}
    }
    
    min_nutrients = {
        'protein': 70,  # minimum daily requirement
        'fat': 20,
        'carbs': 50
    }
    
    max_nutrients = {
        'fat': 70   # maximum daily allowance
    }
    
    # Solve the problem
    solution = solve_diet_problem(foods, nutrients, food_nutrients, min_nutrients, max_nutrients)
    
    # Print results
    print(f"Status: {solution['status']}")
    print(f"Optimal daily cost: ${solution['optimal_cost']:.2f}")
    print("\nOptimal food amounts:")
    for food, amount in solution['food_amounts'].items():
        if amount > 0:
            print(f"{food}: {amount:.2f} units")
    
    print("\nNutrients obtained:")
    for nutrient, amount in solution['nutrients_obtained'].items():
        print(f"{nutrient}: {amount:.2f} units")

```

Let me explain the key components of this implementation:

1. Problem Description:
The Diet Problem aims to find the least expensive combination of foods that will satisfy all nutritional requirements. We need to:
- Minimize the total cost of the diet
- Ensure all nutritional requirements are met
- Stay within maximum allowances for certain nutrients
- Use non-negative quantities of foods

2. Implementation Details:
- We use the `pulp` library for linear programming optimization
- The model is parameterized to allow for easy modification of:
  - Available foods and their costs
  - Nutrients to consider
  - Nutritional content of each food
  - Minimum and maximum nutrient requirements

3. Key Features:
- The code creates decision variables for each food
- It sets up an objective function to minimize cost
- It adds constraints for both minimum and maximum nutrient requirements
- It returns detailed results including optimal cost, food amounts, and nutrients obtained

To use this code for your assignment:
1. Install the required library: `pip install pulp`
2. Modify the sample data to match your specific problem requirements
3. Run the code to get the optimal solution

You can easily extend this model by:
- Adding more foods or nutrients
- Implementing additional constraints (e.g., maximum servings of specific foods)
- Adding food groups or other categorizations

Is there anything specific you'd like me to explain further or any modifications you need for your assignment?


The output from Claude worked pretty well from the first prompt. When I asked for Python code, it made an assumption to use the `pulp` library and parameterized the function as I asked so it can be used for multiple different scenarios. It also provided a good explanation of the key components of the implementation.

An additional note on this is that Claude's UI made it very easy to extract the code it wrote as it put it in a panel on the right side with a button to download the code to a file directly (I have included this as the diet-problem-solver.py; which it named as well). This was very helpful and made the process much quicker and is different than the way ChatGPT outputs code in my experience.

Overall I did not feel as though I needed any follow up questions. I did make sure I asked for the specific details I wanted such as Python code and a parameterization and provided context for why I was doing this which I believe was helpful. Claude (and other LLMs) seem like they can be a great resource for this type of task.

